### PR TITLE
Added new typing / user input UI for the game window.

### PR DESCRIPTION
**Pull request by YingLiang2**
Refactored the component `HighlightText` to have a new typing UI. Now the user can enter characters and have dynamic feedback seeing if they entered a correct or an incorrect character with a caret guiding them. Think of a Monkeytype-Esqe typing UI.

Implementation Details:
* Each character of our game text have been mapped to **a span with a state** therefore we are always updating the state of a span to determine the styling of the character.
* With each character being a span, we have now a position for our caret to be placed which is the left of each span simulating a cursor / caret movement.

This ticket can be further improved and refactored but as this is a first version/iteration I have not refactored too hard. The caret placement/details requires further improvement.

### CODE REVIEW

File: web/src/app/interfaces/CharState.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export default interface CharState {
```
Comment (YingLiang2) [inline web/src/app/interfaces/CharState.ts:1]: When the game text is rendered, each character will have a state assigned to it which is how we determine what colour of text to render.

Currently index is not being used, but may be used in the future.

**Comment by YingLiang2** [comment]
Also @jonathanplacatka why do we have a score (score tracker?). I feel like it doesn't make sense in terms of a typing game no? Doesn't affect things now but I'm just wondering if we need it later.

**Comment by jonathanplacatka** [comment]
The scoreboard (e.g., the column on playerlist) is a placeholder because we don't have an actual race progress indicator right now. The "score" variable is to keep track of how close everyone is to done the race

**Comment by YingLiang2** [comment]
> https://github.com/jonathanplacatka/typing-game/pull/27#issuecomment-2227534197

Maybe we should add the score out of something then such as 5/14 words?

**Review by jonathanplacatka — APPROVED** [review]

**Pull request merged**